File: services/permission-service/src/scope_builder.py
from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any

from policy_loader import load_roles


DEFAULT_FORBIDDEN_TAGS = {
    "visitor": ["internal_pricing", "sales_script", "crm_rule"],
    "student": ["internal_pricing", "sales_script", "crm_rule"],
    "parent": ["internal_pricing", "sales_script", "crm_rule"],
}


def build_scope(identity: dict[str, Any], project_root: Path) -> dict[str, Any]:
    roles = load_roles(project_root)
    role = identity["role"]
    if role not in roles:
        raise ValueError(f"Unknown role: {role}")
    policy = roles[role]
    return {
        "user_id": identity.get("user_id"),
        "role": role,
        "campus": identity.get("campus", "all"),
        "auth_level": identity.get("auth_level", "anonymous"),
        "allowed_visibility": policy.get("allowed_visibility", []),
        "allowed_data_levels": policy.get("allowed_data_levels", []),
        "allowed_roles": _allowed_roles_for(role),
        "forbidden_tags": policy.get("forbidden_tags", DEFAULT_FORBIDDEN_TAGS.get(role, [])),
    }


def _allowed_roles_for(role: str) -> list[str]:
    if role in {"visitor", "student", "parent", "customer"}:
        return ["visitor", "student", "parent", "customer"]
    if role == "sales":
        return ["visitor", "student", "parent", "sales"]
    return [role]


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--root", type=Path, default=Path(__file__).resolve().parents[3])
    parser.add_argument("--identity", required=True, help="JSON identity")
    args = parser.parse_args()
    print(json.dumps(build_scope(json.loads(args.identity), args.root), ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: services/permission-service/ts-src/scopeBuilder.ts
import { loadRoles } from "./policyLoader.js";

const DEFAULT_FORBIDDEN_TAGS: Record<string, string[]> = {
  visitor: ["internal_pricing", "sales_script", "crm_rule"],
  student: ["internal_pricing", "sales_script", "crm_rule"],
  parent: ["internal_pricing", "sales_script", "crm_rule"],
};

function allowedRolesFor(role: string): string[] {
  if (["visitor", "student", "parent", "customer"].includes(role)) {
    return ["visitor", "student", "parent", "customer"];
  }
  if (role === "sales") {
    return ["visitor", "student", "parent", "sales"];
  }
  return [role];
}

export function buildScope(
  identity: Record<string, unknown>,
  projectRoot: string,
): Record<string, unknown> {
  const roles = loadRoles(projectRoot);
  const role = identity.role as string;
  const policy = roles[role] as Record<string, unknown> | undefined;
  if (!policy) {
    throw new Error(`Unknown role: ${role}`);
  }

  return {
    user_id: identity.user_id ?? null,
    role,
    campus: (identity.campus as string) ?? "all",
    auth_level: (identity.auth_level as string) ?? "anonymous",
    allowed_visibility: (policy.allowed_visibility as string[] | undefined) ?? [],
    allowed_data_levels: (policy.allowed_data_levels as string[] | undefined) ?? [],
    allowed_roles: allowedRolesFor(role),
    forbidden_tags:
      (policy.forbidden_tags as string[] | undefined) ??
      DEFAULT_FORBIDDEN_TAGS[role] ??
      [],
  };
}
